src/meta/processors/admin/Balancer.cpp

```diff
@@ -246,106 +253,190 @@ Balancer::genTasks(GraphSpaceID spaceId,
     std::vector<BalanceTask> tasks;
     // 1. Iterate through all hosts that would not be included in newHostParts,
     //    move all parts in them to host with minimum part in newHostParts
-    for (auto& h : hostDel) {
-        auto& lostParts = hostParts[h];
+    for (auto& lostHost : lostHosts) {
+        auto& lostParts = hostParts[lostHost];
         for (auto& partId : lostParts) {
+            LOG(INFO) << "Try balance part " << partId << " for lost host " << lostHost;
             // check whether any peers which is alive
             auto alive = checkReplica(hostParts, activeHosts, spaceReplica, partId);
             if (!alive.ok()) {
-                LOG(ERROR) << "Error: " << alive;
+                LOG(ERROR) << "Check Replica failed: " << alive
+                           << " Part: " << partId;
                 return cpp2::ErrorCode::E_NO_VALID_HOST;
             }
-            // find a host with minimum parts which doesn't have this part
-            auto ret = hostWithMinimalParts(newHostParts, partId);
-            if (!ret.ok()) {
-                LOG(ERROR) << "Error: " << ret.status();
-                return cpp2::ErrorCode::E_NO_VALID_HOST;
+
+            if (transferLostHost(tasks, newHostParts, lostHost,
+                                 spaceId, partId, dependentOnGroup)) {
+                LOG(ERROR) << "";
             }
-            auto& luckyHost = ret.value();
-            newHostParts[luckyHost].emplace_back(partId);
-            tasks.emplace_back(plan_->id_,
-                               spaceId,
-                               partId,
-                               h,
-                               luckyHost,
-                               kv_,
-                               client_);
         }
     }
+
     if (newHostParts.size() < 2) {
         LOG(INFO) << "Too few hosts, no need for balance!";
         return cpp2::ErrorCode::E_NO_VALID_HOST;
     }
     // 2. Make all hosts in newHostParts balanced
-    balanceParts(plan_->id_, spaceId, newHostParts, totalParts, tasks);
-    return tasks;
+    if (balanceParts(plan_->id_, spaceId, newHostParts, totalParts, tasks)) {
+        return tasks;
+    } else {
+        return cpp2::ErrorCode::E_BAD_BALANCE_PLAN;
+    }
+}
+
+bool Balancer::transferLostHost(std::vector<BalanceTask>& tasks,
+                                HostParts& newHostParts,
+                                const HostAddr& source,
+                                GraphSpaceID spaceId,
+                                PartitionID partId,
+                                bool dependentOnGroup) {
+    // find a host with minimum parts which doesn't have this part
+    auto ret = hostWithMinimalParts(newHostParts, partId);
+    if (!ret.ok()) {
+        LOG(ERROR) << "Can't find a host which doesn't have part:" << partId;
+        return false;
+    }
+
+    auto& targetHost = ret.value();
+    if (dependentOnGroup && !checkZoneLegal(source, targetHost, partId)) {
```

Comment:
It seems there are not a big difference with previous version

Reply:
The method is too big and this's what make confuse. Not the function.

PS  some modify will be update later

Reply:
👌